File: .claude/migration-tools/python/abap_analyzer.py
"""
ABAP Code Analyzer for Migration to Python

This module provides utilities to analyze ABAP code and extract information
useful for migration to Python.
"""

import re
from dataclasses import dataclass
from typing import List, Dict, Set, Optional
from enum import Enum


class ComplexityLevel(Enum):
    """Migration complexity levels"""
    SIMPLE = "Simple"
    MEDIUM = "Medium"
    COMPLEX = "Complex"
    VERY_COMPLEX = "Very Complex"


@dataclass
class DataDeclaration:
    """Represents an ABAP data declaration"""
    name: str
    abap_type: str
    is_table: bool = False
    is_structure: bool = False
    python_equivalent: str = ""


@dataclass
class DatabaseOperation:
    """Represents a database operation"""
    operation: str  # SELECT, INSERT, UPDATE, DELETE, MODIFY
    table: str
    fields: List[str]
    conditions: List[str]
    line_number: int


@dataclass
class FunctionModule:
    """Represents an ABAP function module"""
    name: str
    importing: List[DataDeclaration]
    exporting: List[DataDeclaration]
    tables: List[DataDeclaration]
    exceptions: List[str]


@dataclass
class AnalysisResult:
    """Complete analysis result"""
    complexity: ComplexityLevel
    data_declarations: List[DataDeclaration]
    database_operations: List[DatabaseOperation]
    function_calls: List[str]
    class_references: List[str]
    risks: List[str]
    recommendations: List[str]
    estimated_effort_hours: int


class ABAPAnalyzer:
    """Analyze ABAP code for migration"""

    # ABAP to Python type mappings
    TYPE_MAPPING = {
        'i': 'int',
        'int8': 'int',
        'p': 'Decimal',
        'f': 'float',
        'string': 'str',
        'c': 'str',
        'n': 'str',  # Numeric string
        'd': 'date',
        't': 'time',
        'x': 'bytes',
        'xstring': 'bytes',
    }

    # DDIC type mappings
    DDIC_MAPPING = {
        'DATS': 'date',
        'TIMS': 'time',
        'NUMC': 'str',
        'CURR': 'Decimal',
        'QUAN': 'Decimal',
        'CLNT': 'str',
        'LANG': 'str',
        'CUKY': 'str',
        'UNIT': 'str',
    }

    def __init__(self, abap_code: str):
        self.code = abap_code
        self.lines = abap_code.split('\n')

    def analyze(self) -> AnalysisResult:
        """Perform complete analysis"""
        data_declarations = self._extract_data_declarations()
        db_operations = self._extract_database_operations()
        function_calls = self._extract_function_calls()
        class_references = self._extract_class_references()

        complexity = self._assess_complexity(
            data_declarations, db_operations, function_calls
        )

        risks = self._identify_risks()
        recommendations = self._generate_recommendations(complexity)
        effort = self._estimate_effort(complexity, db_operations, function_calls)

        return AnalysisResult(
            complexity=complexity,
            data_declarations=data_declarations,
            database_operations=db_operations,
            function_calls=function_calls,
            class_references=class_references,
            risks=risks,
            recommendations=recommendations,
            estimated_effort_hours=effort
        )

    def _extract_data_declarations(self) -> List[DataDeclaration]:
        """Extract DATA declarations"""
        declarations = []

        # Pattern for DATA declarations
        data_pattern = r'DATA[:\s]+(\w+)\s+TYPE\s+(?:TABLE\s+OF\s+)?(\w+)'

        for line in self.lines:
            line_upper = line.upper()
            if 'DATA' in line_upper and 'TYPE' in line_upper:
                match = re.search(data_pattern, line_upper)
                if match:
                    var_name = match.group(1).lower()
                    abap_type = match.group(2).lower()
                    is_table = 'TABLE OF' in line_upper

                    # Map to Python type
                    python_type = self._map_to_python_type(abap_type, is_table)

                    declarations.append(DataDeclaration(
                        name=var_name,
                        abap_type=abap_type,
                        is_table=is_table,
                        python_equivalent=python_type
                    ))

        return declarations

    def _map_to_python_type(self, abap_type: str, is_table: bool) -> str:
        """Map ABAP type to Python equivalent"""
        base_type = self.TYPE_MAPPING.get(
            abap_type,
            self.DDIC_MAPPING.get(abap_type, 'Any')
        )

        if is_table:
            return f'List[{base_type}]'

        return base_type

    def _extract_database_operations(self) -> List[DatabaseOperation]:
        """Extract database operations"""
        operations = []

        db_keywords = ['SELECT', 'INSERT', 'UPDATE', 'DELETE', 'MODIFY']

        for i, line in enumerate(self.lines):
            line_upper = line.upper().strip()

            for keyword in db_keywords:
                if line_upper.startswith(keyword):
                    # Extract table name
                    table_match = re.search(r'FROM\s+(@?\w+)', line_upper)
                    table = table_match.group(1).replace('@', '') if table_match else 'unknown'

                    # Extract fields (simplified)
                    fields = []
                    if keyword == 'SELECT':
                        fields_match = re.search(r'SELECT\s+(.*?)\s+FROM', line_upper)
                        if fields_match:
                            fields_str = fields_match.group(1)
                            if fields_str != '*':
                                fields = [f.strip() for f in fields_str.split(',')]

                    # Extract WHERE conditions
                    conditions = []
                    if 'WHERE' in line_upper:
                        where_match = re.search(r'WHERE\s+(.*?)(?:\.|$)', line_upper)
                        if where_match:
                            conditions = [where_match.group(1)]

                    operations.append(DatabaseOperation(
                        operation=keyword,
                        table=table,
                        fields=fields,
                        conditions=conditions,
                        line_number=i + 1
                    ))

        return operations

    def _extract_function_calls(self) -> List[str]:
        """Extract function module calls"""
        function_calls = []

        for line in self.lines:
            # CALL FUNCTION
            if 'CALL FUNCTION' in line.upper():
                match = re.search(r"CALL FUNCTION\s+'([^']+)'", line, re.IGNORECASE)
                if match:
                    function_calls.append(match.group(1))

            # Method calls
            if '->' in line:
                match = re.search(r'(\w+)->(\w+)', line)
                if match:
                    function_calls.append(f"{match.group(1)}.{match.group(2)}")

        return function_calls

    def _extract_class_references(self) -> List[str]:
        """Extract class references"""
        classes = set()

        for line in self.lines:
            # CREATE OBJECT
            if 'CREATE OBJECT' in line.upper():
                match = re.search(r'TYPE\s+(\w+)', line, re.IGNORECASE)
                if match:
                    classes.add(match.group(1))

            # NEW operator
            if 'NEW' in line.upper():
                match = re.search(r'NEW\s+(\w+)', line, re.IGNORECASE)
                if match:
                    classes.add(match.group(1))

        return list(classes)

    def _assess_complexity(
        self,
        declarations: List[DataDeclaration],
        db_ops: List[DatabaseOperation],
        function_calls: List[str]
    ) -> ComplexityLevel:
        """Assess migration complexity"""
        score = 0

        # Data complexity
        if len(declarations) > 20:
            score += 2
        elif len(declarations) > 10:
            score += 1

        # Database complexity
        if len(db_ops) > 10:
            score += 2
        elif len(db_ops) > 5:
            score += 1

        # Integration complexity
        if len(function_calls) > 15:
            score += 2
        elif len(function_calls) > 5:
            score += 1

        # Special patterns
        code_upper = self.code.upper()
        if 'CALL FUNCTION' in code_upper and 'DESTINATION' in code_upper:
            score += 2  # RFC calls

        if 'COMMIT WORK' in code_upper or 'ROLLBACK WORK' in code_upper:
            score += 1  # Transaction handling

        if 'AUTHORITY-CHECK' in code_upper:
            score += 1  # Authorization

        if 'BAPI' in code_upper:
            score += 1  # BAPI usage

        # Determine level
        if score == 0:
            return ComplexityLevel.SIMPLE
        elif score <= 3:
            return ComplexityLevel.MEDIUM
        elif score <= 6:
            return ComplexityLevel.COMPLEX
        else:
            return ComplexityLevel.VERY_COMPLEX

    def _identify_risks(self) -> List[str]:
        """Identify migration risks"""
        risks = []
        code_upper = self.code.upper()

        if 'NATIVE SQL' in code_upper or 'EXEC SQL' in code_upper:
            risks.append("Native SQL detected - requires database-specific review")

        if 'CALL FUNCTION' in code_upper and 'DESTINATION' in code_upper:
            risks.append("RFC calls detected - requires integration strategy")

        if 'AUTHORITY-CHECK' in code_upper:
            risks.append("Authorization logic - must map to new security model")

        if 'UPDATE TASK' in code_upper or 'IN UPDATE TASK' in code_upper:
            risks.append("Update tasks - requires async processing design")

        if 'COMMIT WORK AND WAIT' in code_upper:
            risks.append("Synchronous commits - consider async patterns")

        if 'ENHANCEMENT-POINT' in code_upper or 'BADI' in code_upper:
            risks.append("Enhancement framework - needs plugin architecture")

        if len(self._extract_database_operations()) > 20:
            risks.append("Heavy database usage - performance testing required")

        return risks

    def _generate_recommendations(self, complexity: ComplexityLevel) -> List[str]:
        """Generate migration recommendations"""
        recommendations = []

        recommendations.append(f"Complexity Level: {complexity.value}")

        if complexity in [ComplexityLevel.SIMPLE, ComplexityLevel.MEDIUM]:
            recommendations.append("✓ Good candidate for direct migration")
            recommendations.append("Consider using FastAPI for web services")
            recommendations.append("Use SQLAlchemy for database operations")
        else:
            recommendations.append("⚠ Consider phased migration approach")
            recommendations.append("Extract business logic into service layer")
            recommendations.append("Design API contract before implementation")

        if 'SELECT' in self.code.upper():
            recommendations.append("Use ORM (SQLAlchemy) for database access")
            recommendations.append("Add database connection pooling")

        if 'CALL FUNCTION' in self.code.upper():
            recommendations.append("Convert function modules to REST APIs")
            recommendations.append("Use FastAPI or Flask for HTTP endpoints")

        recommendations.append("Implement comprehensive unit tests (pytest)")
        recommendations.append("Add API documentation (OpenAPI/Swagger)")
        recommendations.append("Set up CI/CD pipeline")

        return recommendations

    def _estimate_effort(
        self,
        complexity: ComplexityLevel,
        db_ops: List[DatabaseOperation],
        function_calls: List[str]
    ) -> int:
        """Estimate migration effort in hours"""
        base_hours = {
            ComplexityLevel.SIMPLE: 8,
            ComplexityLevel.MEDIUM: 24,
            ComplexityLevel.COMPLEX: 80,
            ComplexityLevel.VERY_COMPLEX: 200
        }

        hours = base_hours[complexity]

        # Add for database operations
        hours += len(db_ops) * 2

        # Add for function calls
        hours += len(function_calls) * 3

        # Add for testing and documentation
        hours += hours * 0.3

        return int(hours)


def analyze_abap_code(abap_code: str) -> AnalysisResult:
    """Convenience function to analyze ABAP code"""
    analyzer = ABAPAnalyzer(abap_code)
    return analyzer.analyze()


def print_analysis_report(result: AnalysisResult):
    """Print formatted analysis report"""
    print("=" * 80)
    print("ABAP Migration Analysis Report")
    print("=" * 80)

    print(f"\n📊 Complexity: {result.complexity.value}")
    print(f"⏱️  Estimated Effort: {result.estimated_effort_hours} hours")

    print(f"\n📋 Data Declarations ({len(result.data_declarations)}):")
    for decl in result.data_declarations[:10]:  # Show first 10
        print(f"  • {decl.name}: {decl.abap_type} → {decl.python_equivalent}")
    if len(result.data_declarations) > 10:
        print(f"  ... and {len(result.data_declarations) - 10} more")

    print(f"\n💾 Database Operations ({len(result.database_operations)}):")
    for op in result.database_operations[:10]:
        print(f"  • {op.operation} on {op.table} (line {op.line_number})")
    if len(result.database_operations) > 10:
        print(f"  ... and {len(result.database_operations) - 10} more")

    print(f"\n📞 Function Calls ({len(result.function_calls)}):")
    for call in result.function_calls[:10]:
        print(f"  • {call}")
    if len(result.function_calls) > 10:
        print(f"  ... and {len(result.function_calls) - 10} more")

    if result.risks:
        print(f"\n⚠️  Risks ({len(result.risks)}):")
        for risk in result.risks:
            print(f"  • {risk}")

    print(f"\n💡 Recommendations ({len(result.recommendations)}):")
    for rec in result.recommendations:
        print(f"  • {rec}")

    print("\n" + "=" * 80)


if __name__ == "__main__":
    # Example usage
    sample_abap = """
    REPORT z_test_migration.

    DATA: lt_customers TYPE TABLE OF kna1,
          lv_count TYPE i.

    SELECT * FROM kna1
      INTO TABLE @lt_customers
      WHERE land1 = 'US'.

    lv_count = lines( lt_customers ).

    LOOP AT lt_customers INTO DATA(ls_customer).
      WRITE: / ls_customer-kunnr, ls_customer-name1.
    ENDLOOP.
    """

    result = analyze_abap_code(sample_abap)
    print_analysis_report(result)

File: .claude/migration-tools/javascript/abapAnalyzer.ts
/**
 * ABAP Code Analyzer for Migration to JavaScript/TypeScript
 *
 * Analyzes ABAP code and extracts information useful for migration
 */

export enum ComplexityLevel {
  SIMPLE = 'Simple',
  MEDIUM = 'Medium',
  COMPLEX = 'Complex',
  VERY_COMPLEX = 'Very Complex'
}

export interface DataDeclaration {
  name: string;
  abapType: string;
  isTable: boolean;
  isStructure: boolean;
  typeScriptEquivalent: string;
}

export interface DatabaseOperation {
  operation: 'SELECT' | 'INSERT' | 'UPDATE' | 'DELETE' | 'MODIFY';
  table: string;
  fields: string[];
  conditions: string[];
  lineNumber: number;
}

export interface FunctionModule {
  name: string;
  importing: DataDeclaration[];
  exporting: DataDeclaration[];
  tables: DataDeclaration[];
  exceptions: string[];
}

export interface AnalysisResult {
  complexity: ComplexityLevel;
  dataDeclarations: DataDeclaration[];
  databaseOperations: DatabaseOperation[];
  functionCalls: string[];
  classReferences: string[];
  risks: string[];
  recommendations: string[];
  estimatedEffortHours: number;
}

export class ABAPAnalyzer {
  private code: string;
  private lines: string[];

  // ABAP to TypeScript type mappings
  private static readonly TYPE_MAPPING: Record<string, string> = {
    'i': 'number',
    'int8': 'bigint',
    'p': 'number',  // Use Decimal library for precision
    'f': 'number',
    'string': 'string',
    'c': 'string',
    'n': 'string',
    'd': 'Date',
    't': 'Date',
    'x': 'Buffer',
    'xstring': 'Buffer',
  };

  // DDIC type mappings
  private static readonly DDIC_MAPPING: Record<string, string> = {
    'DATS': 'Date',
    'TIMS': 'Date',
    'NUMC': 'string',
    'CURR': 'number',
    'QUAN': 'number',
    'CLNT': 'string',
    'LANG': 'string',
    'CUKY': 'string',
    'UNIT': 'string',
  };

  constructor(abapCode: string) {
    this.code = abapCode;
    this.lines = abapCode.split('\n');
  }

  public analyze(): AnalysisResult {
    const dataDeclarations = this.extractDataDeclarations();
    const databaseOperations = this.extractDatabaseOperations();
    const functionCalls = this.extractFunctionCalls();
    const classReferences = this.extractClassReferences();

    const complexity = this.assessComplexity(
      dataDeclarations,
      databaseOperations,
      functionCalls
    );

    const risks = this.identifyRisks();
    const recommendations = this.generateRecommendations(complexity);
    const effort = this.estimateEffort(complexity, databaseOperations, functionCalls);

    return {
      complexity,
      dataDeclarations,
      databaseOperations,
      functionCalls,
      classReferences,
      risks,
      recommendations,
      estimatedEffortHours: effort
    };
  }

  private extractDataDeclarations(): DataDeclaration[] {
    const declarations: DataDeclaration[] = [];
    const dataPattern = /DATA[:\s]+(\w+)\s+TYPE\s+(?:TABLE\s+OF\s+)?(\w+)/i;

    for (const line of this.lines) {
      const lineUpper = line.toUpperCase();
      if (lineUpper.includes('DATA') && lineUpper.includes('TYPE')) {
        const match = lineUpper.match(dataPattern);
        if (match) {
          const varName = match[1].toLowerCase();
          const abapType = match[2].toLowerCase();
          const isTable = lineUpper.includes('TABLE OF');

          const tsType = this.mapToTypeScriptType(abapType, isTable);

          declarations.push({
            name: varName,
            abapType,
            isTable,
            isStructure: false,
            typeScriptEquivalent: tsType
          });
        }
      }
    }

    return declarations;
  }

  private mapToTypeScriptType(abapType: string, isTable: boolean): string {
    const baseType = ABAPAnalyzer.TYPE_MAPPING[abapType] ||
                     ABAPAnalyzer.DDIC_MAPPING[abapType] ||
                     'any';

    return isTable ? `${baseType}[]` : baseType;
  }

  private extractDatabaseOperations(): DatabaseOperation[] {
    const operations: DatabaseOperation[] = [];
    const dbKeywords = ['SELECT', 'INSERT', 'UPDATE', 'DELETE', 'MODIFY'];

    this.lines.forEach((line, index) => {
      const lineUpper = line.toUpperCase().trim();

      for (const keyword of dbKeywords) {
        if (lineUpper.startsWith(keyword)) {
          // Extract table name
          const tableMatch = lineUpper.match(/FROM\s+(@?\w+)/);
          const table = tableMatch ? tableMatch[1].replace('@', '') : 'unknown';

          // Extract fields
          const fields: string[] = [];
          if (keyword === 'SELECT') {
            const fieldsMatch = lineUpper.match(/SELECT\s+(.*?)\s+FROM/);
            if (fieldsMatch && fieldsMatch[1] !== '*') {
              fields.push(...fieldsMatch[1].split(',').map(f => f.trim()));
            }
          }

          // Extract WHERE conditions
          const conditions: string[] = [];
          if (lineUpper.includes('WHERE')) {
            const whereMatch = lineUpper.match(/WHERE\s+(.*?)(?:\.|$)/);
            if (whereMatch) {
              conditions.push(whereMatch[1]);
            }
          }

          operations.push({
            operation: keyword as any,
            table,
            fields,
            conditions,
            lineNumber: index + 1
          });
        }
      }
    });

    return operations;
  }

  private extractFunctionCalls(): string[] {
    const functionCalls: string[] = [];

    for (const line of this.lines) {
      // CALL FUNCTION
      if (line.toUpperCase().includes('CALL FUNCTION')) {
        const match = line.match(/CALL FUNCTION\s+'([^']+)'/i);
        if (match) {
          functionCalls.push(match[1]);
        }
      }

      // Method calls
      if (line.includes('->')) {
        const match = line.match(/(\w+)->(\w+)/);
        if (match) {
          functionCalls.push(`${match[1]}.${match[2]}`);
        }
      }
    }

    return functionCalls;
  }

  private extractClassReferences(): string[] {
    const classes = new Set<string>();

    for (const line of this.lines) {
      // CREATE OBJECT
      if (line.toUpperCase().includes('CREATE OBJECT')) {
        const match = line.match(/TYPE\s+(\w+)/i);
        if (match) {
          classes.add(match[1]);
        }
      }

      // NEW operator
      if (line.toUpperCase().includes('NEW')) {
        const match = line.match(/NEW\s+(\w+)/i);
        if (match) {
          classes.add(match[1]);
        }
      }
    }

    return Array.from(classes);
  }

  private assessComplexity(
    declarations: DataDeclaration[],
    dbOps: DatabaseOperation[],
    functionCalls: string[]
  ): ComplexityLevel {
    let score = 0;

    // Data complexity
    if (declarations.length > 20) score += 2;
    else if (declarations.length > 10) score += 1;

    // Database complexity
    if (dbOps.length > 10) score += 2;
    else if (dbOps.length > 5) score += 1;

    // Integration complexity
    if (functionCalls.length > 15) score += 2;
    else if (functionCalls.length > 5) score += 1;

    // Special patterns
    const codeUpper = this.code.toUpperCase();
    if (codeUpper.includes('CALL FUNCTION') && codeUpper.includes('DESTINATION')) {
      score += 2; // RFC calls
    }

    if (codeUpper.includes('COMMIT WORK') || codeUpper.includes('ROLLBACK WORK')) {
      score += 1; // Transaction handling
    }

    if (codeUpper.includes('AUTHORITY-CHECK')) {
      score += 1; // Authorization
    }

    // Determine level
    if (score === 0) return ComplexityLevel.SIMPLE;
    if (score <= 3) return ComplexityLevel.MEDIUM;
    if (score <= 6) return ComplexityLevel.COMPLEX;
    return ComplexityLevel.VERY_COMPLEX;
  }

  private identifyRisks(): string[] {
    const risks: string[] = [];
    const codeUpper = this.code.toUpperCase();

    if (codeUpper.includes('NATIVE SQL') || codeUpper.includes('EXEC SQL')) {
      risks.push('Native SQL detected - requires database-specific review');
    }

    if (codeUpper.includes('CALL FUNCTION') && codeUpper.includes('DESTINATION')) {
      risks.push('RFC calls detected - requires integration strategy');
    }

    if (codeUpper.includes('AUTHORITY-CHECK')) {
      risks.push('Authorization logic - must map to new security model');
    }

    if (codeUpper.includes('UPDATE TASK') || codeUpper.includes('IN UPDATE TASK')) {
      risks.push('Update tasks - requires async processing design');
    }

    if (codeUpper.includes('COMMIT WORK AND WAIT')) {
      risks.push('Synchronous commits - consider async patterns');
    }

    if (this.databaseOperations.length > 20) {
      risks.push('Heavy database usage - performance testing required');
    }

    return risks;
  }

  private generateRecommendations(complexity: ComplexityLevel): string[] {
    const recommendations: string[] = [];

    recommendations.push(`Complexity Level: ${complexity}`);

    if (complexity === ComplexityLevel.SIMPLE || complexity === ComplexityLevel.MEDIUM) {
      recommendations.push('✓ Good candidate for direct migration');
      recommendations.push('Consider using NestJS for enterprise-grade structure');
      recommendations.push('Use Prisma or TypeORM for database operations');
    } else {
      recommendations.push('⚠ Consider phased migration approach');
      recommendations.push('Extract business logic into service layer');
      recommendations.push('Design API contract before implementation');
    }

    if (this.code.toUpperCase().includes('SELECT')) {
      recommendations.push('Use ORM (Prisma/TypeORM) for database access');
      recommendations.push('Implement connection pooling');
    }

    if (this.code.toUpperCase().includes('CALL FUNCTION')) {
      recommendations.push('Convert function modules to REST/GraphQL APIs');
      recommendations.push('Use NestJS with decorators for clean architecture');
    }

    recommendations.push('Implement comprehensive tests (Jest/Vitest)');
    recommendations.push('Add API documentation (Swagger/OpenAPI)');
    recommendations.push('Set up TypeScript strict mode');
    recommendations.push('Use Zod for runtime validation');

    return recommendations;
  }

  private estimateEffort(
    complexity: ComplexityLevel,
    dbOps: DatabaseOperation[],
    functionCalls: string[]
  ): number {
    const baseHours: Record<ComplexityLevel, number> = {
      [ComplexityLevel.SIMPLE]: 8,
      [ComplexityLevel.MEDIUM]: 24,
      [ComplexityLevel.COMPLEX]: 80,
      [ComplexityLevel.VERY_COMPLEX]: 200
    };

    let hours = baseHours[complexity];

    // Add for database operations
    hours += dbOps.length * 2;

    // Add for function calls
    hours += functionCalls.length * 3;

    // Add for testing and documentation
    hours += hours * 0.3;

    return Math.round(hours);
  }
}

export function analyzeAbapCode(abapCode: string): AnalysisResult {
  const analyzer = new ABAPAnalyzer(abapCode);
  return analyzer.analyze();
}

export function printAnalysisReport(result: AnalysisResult): void {
  console.log('='.repeat(80));
  console.log('ABAP Migration Analysis Report');
  console.log('='.repeat(80));

  console.log(`\n📊 Complexity: ${result.complexity}`);
  console.log(`⏱️  Estimated Effort: ${result.estimatedEffortHours} hours`);

  console.log(`\n📋 Data Declarations (${result.dataDeclarations.length}):`);
  result.dataDeclarations.slice(0, 10).forEach(decl => {
    console.log(`  • ${decl.name}: ${decl.abapType} → ${decl.typeScriptEquivalent}`);
  });
  if (result.dataDeclarations.length > 10) {
    console.log(`  ... and ${result.dataDeclarations.length - 10} more`);
  }

  console.log(`\n💾 Database Operations (${result.databaseOperations.length}):`);
  result.databaseOperations.slice(0, 10).forEach(op => {
    console.log(`  • ${op.operation} on ${op.table} (line ${op.lineNumber})`);
  });
  if (result.databaseOperations.length > 10) {
    console.log(`  ... and ${result.databaseOperations.length - 10} more`);
  }

  console.log(`\n📞 Function Calls (${result.functionCalls.length}):`);
  result.functionCalls.slice(0, 10).forEach(call => {
    console.log(`  • ${call}`);
  });
  if (result.functionCalls.length > 10) {
    console.log(`  ... and ${result.functionCalls.length - 10} more`);
  }

  if (result.risks.length > 0) {
    console.log(`\n⚠️  Risks (${result.risks.length}):`);
    result.risks.forEach(risk => {
      console.log(`  • ${risk}`);
    });
  }

  console.log(`\n💡 Recommendations (${result.recommendations.length}):`);
  result.recommendations.forEach(rec => {
    console.log(`  • ${rec}`);
  });

  console.log('\n' + '='.repeat(80));
}

// Example usage
if (require.main === module) {
  const sampleAbap = `
    REPORT z_test_migration.

    DATA: lt_customers TYPE TABLE OF kna1,
          lv_count TYPE i.

    SELECT * FROM kna1
      INTO TABLE @lt_customers
      WHERE land1 = 'US'.

    lv_count = lines( lt_customers ).

    LOOP AT lt_customers INTO DATA(ls_customer).
      WRITE: / ls_customer-kunnr, ls_customer-name1.
    ENDLOOP.
  `;

  const result = analyzeAbapCode(sampleAbap);
  printAnalysisReport(result);
}
